### PR TITLE
Add version 13 support

### LINKED ISSUE #17: Make parser understand YAML diagnostics for TAP version 13
TAP version 13 added optional YAML blocks to tests. Everything else about the spec stayed the same. If the TAP parser was updated to support YAML blocks, then tappy could be considered version 13 "compliant" (even though YAML blocks never had any formally defined schema).

### PR DESCRIPTION
**Pull request by arewm**
Here is my initial implementation of version 13 support to resolve #17. I had to do some trickery in the form of iterator re-generation. As always, let me know if you think something should be approached differently.

I added optional dependencies on `pyyaml` and `more-itertools` which need to be included somewhere in the package. Do you want to use an [`optional-requirements.txt`](https://stackoverflow.com/questions/3664478/optional-dependencies-in-a-pip-requirements-file) file for these?

### CODE REVIEW

**Comment by arewm** [comment]
What is AppVeyor? I do not know why `ModuleNotFoundError` is not defined, isn't that part of the standard library?

File: tap/tests/test_loader.py [hunk 1]
```diff
@@ -116,8 +117,8 @@ def test_skip_plan_aborts_loading(self):
         self.assertEqual(
             'Skipping this test file.', suite._tests[0]._line.description)
 
-    @mock.patch('tap.parser.sys')
-    def test_loads_from_stream(self, mock_sys):
+    @mock.patch('tap.parser.sys.stdin', StringIO('\n'))
```
Comment (arewm) [inline tap/tests/test_loader.py]: This is invalid in Python 2.7. How can I make this work in both 2.7 and 3.x?

File: tap/tests/test_parser.py [hunk 2]
```diff
@@ -196,18 +197,83 @@ def test_parses_file(self):
...
-    @mock.patch('tap.parser.sys')
-    def test_parses_stdin(self, mock_sys):
-        mock_sys.stdin.__iter__.return_value = iter([
-            '1..2\n',
-            'ok 1 A passing test\n',
-            'not ok 2 A failing test\n',
-        ])
-        mock_sys.stdin.__enter__.return_value = None
-        mock_sys.stdin.__exit__.return_value = None
+    def test_parses_yaml(self):
```
Comment (arewm) [inline tap/tests/test_parser.py]: These tests for optional functionality are failing. How can I test both regression (i.e. behaviors with modules not installed) and new functionality?

**Comment by arewm** [comment]
All tests are finally passing. Code coverage reduction is due to inability to test the import statement failing.

**Comment by arewm** [comment]
FYI, this implementation does break the concept of parsing output "line-by-line", however, I do not think there is any way to get around it. When parsing a result line, if there is a following yaml block, all of those lines will be parsed.

I do not know if this change needs to be documented anywhere.

File: .travis.yml [hunk 3]
```diff
@@ -24,6 +24,12 @@ matrix:
         - brew upgrade python
         - python3 -m venv venv
         - source venv/bin/activate
+    - os: linux
```
[the remark this answers is not in the document]
Comment (arewm) [reply .travis.yml]: Sounds good.

File: tap/line.py [hunk 4]
```diff
@@ -16,7 +16,7 @@ class Result(Line):
 
     def __init__(
             self, ok, number=None, description='', directive=None,
-            diagnostics=None):
+            diagnostics=None, yaml=None):
```
[the remark this answers is not in the document]
Comment (arewm) [reply tap/line.py]: Makes sense.

File: tap/parser.py [hunk 5]
```diff
@@ -1,13 +1,27 @@
...
 from tap.line import Bail, Diagnostic, Plan, Result, Unknown, Version
 
+try:
+    from more_itertools import peekable
+    import yaml
+    SUPPORTS_PEEKING = True
+except ImportError:  # pragma: no cover
+    SUPPORTS_PEEKING = False
+    print("""
+        WARNING: Optional imports not found, TAP 13 output will be ignored. To
```
[the remark this answers is not in the document]
Comment (arewm) [reply tap/parser.py]: Yeah, I thought about the noise, but did not think about locating this at the detection of version 13.

I will work on that documentation too.

File: tap/parser.py [hunk 6]
```diff
@@ -73,15 +93,25 @@ def parse(self, fh):
         stripped from the input lines.
         """
         with fh:
-            for line in fh:
-                yield self.parse_line(line.rstrip())
-
-    def parse_line(self, text):
+            first_line = next(fh)
```
[the remark this answers is not in the document]
Comment (arewm) [reply tap/parser.py]: For sure. Too bad this case was not in the test suite. 😉

File: tap/parser.py [hunk 7]
```diff
@@ -73,15 +93,25 @@ def parse(self, fh):
...
+            first_parsed = self.parse_line(first_line.rstrip())
+            if SUPPORTS_PEEKING and first_parsed.category == 'version' and \
+                    int(first_parsed.version) == 13:
+                fh_new = peekable(itertools.chain([first_line], fh))
+                self._try_peeking = True
+            else:
+                fh_new = itertools.chain([first_line], fh)
```
[the remark this answers is not in the document]
Comment (arewm) [reply tap/parser.py]: Not a problem. I was struggling to figure out how to implement this until I figured out the chaining. I do not ~play around with~ utilize itertools enough.

File: tap/parser.py [hunk 8]
```diff
@@ -73,15 +93,25 @@ def parse(self, fh):
...
+                yield self.parse_line(line.rstrip(), fh_new)
+
+    def parse_line(self, text, fh=None):
         """Parse a line into whatever TAP category it belongs."""
         match = self.ok.match(text)
         if match:
-            return self._parse_result(True, match)
+            return self._parse_result(True, match, fh)
 
+        self._reset_last_result = True
```
[the remark this answers is not in the document]
Comment (arewm) [reply tap/parser.py]: Must be

File: tap/parser.py [hunk 9]
```diff
@@ -114,11 +144,39 @@ def _parse_plan(self, match):
...
+            pass
+        concat_yaml = '\n'.join(raw_yaml)
         return Result(
-            ok, match.group('number'), match.group('description').strip(),
-            Directive(match.group('directive')))
+            ok,
+            number=match.group('number'),
+            description=match.group('description').strip(),
+            directive=Directive(match.group('directive')),
+            yaml=yaml.load(concat_yaml)
```
[the remark this answers is not in the document]
Comment (arewm) [reply tap/parser.py]: Good call, this change will make `flake8` complain when I import `yaml` in `parser` and I don't end up using it. I want to ensure that the dependency is present before parsing it.

File: tap/tests/test_parser.py [hunk 10]
```diff
@@ -196,18 +197,171 @@ def test_parses_file(self):
...
+        self.assertIsNone(lines[2].yaml_block)
+        for l in list(range(3, 5)):
+            self.assertEqual('unknown', lines[l].category)
+        self.assertEqual('test', lines[5].category)
+
+    @mock.patch('tap.parser.sys.stdin',
+                StringIO(u"""1..2
+ok 1 A passing test
+not ok 2 A failing test"""))
+    def test_parses_stdin(self):
```
[the remark this answers is not in the document]
Comment (arewm) [reply tap/tests/test_parser.py]: I can do this. The warning will have to be printed during the lazy-loading.

File: docs/consumers.rst [hunk 11]
```diff
@@ -112,6 +112,23 @@ The API specifics are listed below.
...
+be consumed, but any yaml blocks will be parsed as :class:`tap.line.Unknown`. If a 
+:class:`tap.line.Result` object has an associated yaml block, :attr:`~tap.line.Result.yaml_block`
+will return the block converted to a `dict`. Otherwise, it will return `None`.
+
+``tappy`` provides a strict interpretation of the specification. A yaml block will
+only be associated with a result if it immediately follows that result. Any 
+:class:`diagnostic <tap.line.Diagnostic>` between a :class:`result <tap.line.Result>` and a yaml
+block will result in the block lines being parsed as :class:`tap.line.Unknown`.
+
 Line Categories
```
Comment (arewm) [inline docs/consumers.rst]: My local documentation build was not picking up the `yaml_block` property. I assume that is just with me, but can you make sure that it is working for you?

**Comment by arewm** [comment]
Sounds good @mblayman. I will be happy to look at the final merge. I updated the version but did not put in a release date.

FYI, I despise unicode issues.

**Comment by mblayman** [comment]
@arewm sorry for the long delay. Somehow I missed that you addressed issues. I'll give this another look when I have time. It will probably be some time next week when I'm at PyCon.

**Pull request merged**